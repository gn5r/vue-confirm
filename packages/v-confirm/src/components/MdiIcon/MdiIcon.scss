@use "../../styles/index.scss" as styles;

@include styles.theme(mdi-icon) using ($theme) {
  color: map-get($theme, "text-color");
}

.mdi-icon {
  display: flex;
  align-items: center;
  background-color: transparent;
  border: none;
  cursor: pointer;
  outline: none;
  padding: 0;
  appearance: none;
}
