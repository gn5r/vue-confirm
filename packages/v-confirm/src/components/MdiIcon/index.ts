export { default as MdiIcon } from "./MdiIcon";
