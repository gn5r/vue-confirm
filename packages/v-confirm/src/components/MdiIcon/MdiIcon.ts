// Types
import { VNode } from "vue";

// Style
import "./MdiIcon.scss";

// Helpers
import { getSize } from "../../utils/helper";
import { setTextColor } from "../../utils/colorUtil";

// Mixins
import Themeable from "../../mixins/themeable";

export default Themeable.extend({
  name: "icon",
  props: {
    size: {
      type: [Number, String],
      default: 24,
    },
    color: {
      type: String,
      default: "inherit",
    },
  },
  computed: {
    hasClickListener(): boolean {
      return Boolean(this.$listeners.click);
    },
    classes(): object {
      return {
        "mdi-icon": true,
        ...this.themeClasses,
      };
    },
  },
  render(h): VNode {
    let iconName = "";
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    if (this.$slots.default) iconName = this.$slots.default[0].text!.trim();
    const fontSize = getSize(this.size);
    const data = {
      attrs: {
        "area-hidden": !this.hasClickListener,
        type: this.hasClickListener ? "button" : undefined,
      },
      class: this.classes,
      style: {
        width: `${fontSize}px`,
        height: `${fontSize + 12}px`,
      },
      on: this.$listeners,
    };
    return h(
      this.hasClickListener ? "button" : "span",
      setTextColor(this.color, data),
      [
        h(
          "svg",
          {
            attrs: {
              xmlns: "http://www.w3.org/2000/svg",
              viewBox: "0 0 24 24",
              role: "img",
              "area-hidden": true,
            },
            style: {
              "font-size": `${fontSize}px`,
              width: `${fontSize}px`,
              height: `${fontSize}px`,
              fill: "currentColor",
              "vertical-align": "middle",
            },
          },
          [
            h("path", {
              attrs: { d: iconName },
            }),
          ]
        ),
      ]
    );
  },
});
