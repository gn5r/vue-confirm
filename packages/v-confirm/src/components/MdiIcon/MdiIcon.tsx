import { defineComponent, toRef, computed } from "vue";
import type { ComputedRef } from "vue";

import { getSize } from "@/utils/helper";

import { useThemeProps, useThemeClasses } from "@/composables/theme";
import { useTextColor } from "@/composables/color";

import "./MdiIcon.scss";

export const MdiIcon = defineComponent({
  name: "mdi-icon",
  props: {
    size: {
      type: [Number, String],
      default: 24,
    },
    color: {
      type: String,
      default: "inherit",
    },
    ...useThemeProps(),
  },
  emits: ["click"],
  setup(props, { slots, emit, attrs }) {
    let slotIcon: ComputedRef<string | undefined> | undefined;
    if (slots.default) {
      slotIcon = computed(() => {
        const slot = slots.default?.();
        if (!slot) return "";
        return slot[0].children as string;
      });
    }
    const themeClasses = useThemeClasses(props);
    const fontSize = getSize(props.size);
    const { textColorClasses, textColorStyles } = useTextColor(
      toRef(props, "color")
    );

    return () => (
      <button
        type="button"
        class={["mdi-icon", themeClasses.value, textColorClasses.value]}
        style={[
          { width: `${fontSize}px`, height: `${fontSize + 12}px` },
          textColorStyles.value,
        ]}
        onClick={(e: Event) => {
          emit("click", e);
        }}
        {...attrs}
      >
        <svg
          xmlns="http://www.w3.org/2000/svg"
          viewBox="0 0 24 24"
          role="img"
          aria-hidden="true"
          style={{
            "font-size": `${fontSize}px`,
            width: `${fontSize}px`,
            height: `${fontSize}px`,
            fill: "currentColor",
            "vertical-align": "middle",
          }}
        >
          <path d={slotIcon?.value} />
        </svg>
      </button>
    );
  },
});

export type MdiIcon = InstanceType<typeof MdiIcon>;
