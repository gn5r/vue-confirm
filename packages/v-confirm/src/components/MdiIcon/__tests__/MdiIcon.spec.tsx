import { mount } from "@vue/test-utils";
import { describe, it, expect, vi } from "vitest";
import { MdiIcon } from "../MdiIcon";
import { mdiClose } from "@mdi/js";

describe("MdiIcon.tsx", () => {
  it("render a component", () => {
    const wrapper = mount(MdiIcon, { slots: { default: mdiClose } });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("render a dark component", () => {
    const wrapper = mount(MdiIcon, {
      props: { dark: true },
      slots: { default: mdiClose },
    });
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find("button.dark").exists()).toBe(true);
  });

  it("should emitted click event when clicked this component", async () => {
    const func = vi.fn();
    const wrapper = mount(MdiIcon, {
      slots: { default: mdiClose },
      attrs: { onClick: func },
    });
    await wrapper.find("button").trigger("click");
    expect(func).toHaveBeenCalledTimes(1);
    expect(wrapper.emitted("click")).toBeTruthy();
  });
});
