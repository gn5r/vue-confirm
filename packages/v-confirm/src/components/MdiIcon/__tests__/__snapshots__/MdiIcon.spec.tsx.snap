// Vitest Snapshot v1

exports[`MdiIcon.ts > render a component 1`] = `
"<button type=\\"button\\" class=\\"mdi-icon light\\" style=\\"width: 24px; height: 24px; font-size: 24px; caret-color: inherit;\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\" role=\\"img\\" aria-hidden=\\"true\\" style=\\"fill: currentColor;\\">
    <path d=\\"M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z\\"></path>
  </svg></button>"
`;

exports[`MdiIcon.ts > render a dark component 1`] = `
"<button type=\\"button\\" class=\\"mdi-icon dark\\" style=\\"width: 24px; height: 24px; font-size: 24px; caret-color: inherit;\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\" role=\\"img\\" aria-hidden=\\"true\\" style=\\"fill: currentColor;\\">
    <path d=\\"M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z\\"></path>
  </svg></button>"
`;
